### PR TITLE
[16.0] base_tier_validation: migration to 16

### PR DESCRIPTION
**Pull request by bizzappdev**

### CODE REVIEW

**Comment by celm1990** [comment]
Any plan to improve performance??
please see https://github.com/OCA/server-ux/issues/496

**Comment by bizzappdev** [comment]
> Any plan to improve performance?? please see #496

Not in this MR. 
Prio is the Migration first and then might be new MR for improvement of performance.

**Comment by Bart-dh** [comment]
#534

**Comment by StefanRijnhart** [comment]
@bizzappdev can you tell us what still needs to be done?

**Comment by bizzappdev** [comment]
> @bizzappdev can you tell us what still needs to be done?

only testing from the QA team was remaining. now I have confirmation from the QA team also.

it is ready to review

**Comment by StefanRijnhart** [comment]
@bizzappdev thanks! In that case, please check the failing pre-commit linting.

File: base_tier_validation/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# Copyright 2017-19 ForgeFlow S.L. (https://www.forgeflow.com)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+{
+    "name": "Base Tier Validation",
+    "summary": "Implement a validation process based on tiers.",
+    "version": "16.0.0.0.0",
```
Comment (StefanRijnhart) [inline base_tier_validation/__manifest__.py]: As per convention
```suggestion
    "version": "16.0.1.0.0",
```

**Comment by StefanRijnhart** [comment]
@bizzappdev do you know of any of the OCA tier validation implementation modules for 16.0 to test this one with? A WIP branch or PR perhaps?

**Comment by bizzappdev** [comment]
> @bizzappdev do you know of any of the OCA tier validation implementation modules for 16.0 to test this one with? A WIP branch or PR perhaps?

@StefanRijnhart we tested sale_tier_validation from v15. we still have to prepare the MR for that. also, automated test cases worked properly.

**Comment by StefanRijnhart** [comment]
Please check the js linting
```
 error  Parsing error: 'import' and 'export' may appear only with 'sourceType: module'
 ```

**Comment by bizzappdev** [comment]
> Please check the js linting
> 
> ```
>  error  Parsing error: 'import' and 'export' may appear only with 'sourceType: module'
> ```

I really do not have much idea about how to fix this. as in the entire odoo also they have imported like this only. as some suggestions are changing .eslintrc file. but that will be needed at 16.0 level.

File: base_tier_validation/views/res_config_settings_views.xml [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            <xpath expr="//div[@id='companies']" position="inside">
+                <div id="base_tier_validation">
+                    <h2>Tier Validation</h2>
+                    <div
+                        class="row mt16 o_settings_container"
+                        name="base_tier_validation_option_setting_container"
+                    >
+                        <div
+                            class="col-12 col-lg-6 o_setting_box"
+                            id="module_base_tier_validation_formula"
```
Comment (StefanRijnhart) [inline base_tier_validation/views/res_config_settings_views.xml:20]: All these modules are not yet migrated, and some of them have not even been migrated to 15 yet, so offering these options here seems misleading. Maybe these options should be commented out here and reenabled in each of these modules' migration PRs?

**Comment by StefanRijnhart** [comment]
> I really do not have much idea about how to fix this. as in the entire odoo also they have imported like this only. as some suggestions are changing .eslintrc file. but that will be needed at 16.0 level.

You can fix the errors by renaming the .js files to .esm.js files as mentioned in the 15.0 migration issue (https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-15.0#tasks-to-do-in-the-migration). This will trigger further eslint messages that need to be followed up.

**Review by StefanRijnhart — CHANGES_REQUESTED** [review]
Thanks for the huge work migrating this module and all its javascript! I tested the module successfully. The code still needs some work on the JS linting.

**Comment by rven** [comment]
``> Thanks for the huge work migrating this module and all its javascript! I tested the module successfully. The code still needs some work on the JS linting.

@StefanRijnhart
We discovered the same problem with the linting

I think the solution for the JS linting is the following:
```yml
parserOptions:
  ecmaVersion: 2019

overrides:
  - files:
      - "**/*.esm.js"
    parserOptions:
      sourceType: module
```
Should be adapted to
```yml
parserOptions:
  ecmaVersion: 2019
  sourceType: module

overrides:
  - files:
      - "**/*.esm.js"
    parserOptions:
      sourceType: module
```
After this change, the linting error will dissapear.

**Comment by pedrobaeza** [comment]
JS files should be renamed to `.esm.js` if they use the new ES6 syntax.

**Comment by bizzappdev** [comment]
> > I really do not have much idea about how to fix this. as in the entire odoo also they have imported like this only. as some suggestions are changing .eslintrc file. but that will be needed at 16.0 level.
> 
> You can fix the errors by renaming the .js files to .esm.js files as mentioned in the 15.0 migration issue (https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-15.0#tasks-to-do-in-the-migration). This will trigger further eslint messages that need to be followed up.

Pre-commits are fixed. you can review again.
code coverage is already more than 90%. should we increase the coverage?

**Comment by bizzappdev** [comment]
Pre-commits are fixed. you can review it again.
code coverage is already more than 90%. should we increase the coverage?
@StefanRijnhart

File: base_tier_validation/__manifest__.py (hunk 1 above)
Comment (StefanRijnhart) [reply base_tier_validation/__manifest__.py]: Would you consider changing this?

**Comment by StefanRijnhart** [comment]
Thanks for the js updates! Looking good. Please check my version number comment. It's always good to improve coverage, but it's going to be hard to make that 96% mark. I had a look at what's still missing coverage, and I don't see anything critical.

File: base_tier_validation/__manifest__.py (hunk 1 above)
Comment (bizzappdev) [reply base_tier_validation/__manifest__.py]: done

**Review by StefanRijnhart — APPROVED** [review]
Thanks! Tested + code reviewed.

**Review by alan196 — APPROVED** [review]
Tested functionally 👍

**Comment by alan196** [comment]
I tested the module in a custom model, and I get an issue when I try to click on the notification in the tray menu.

It looks that is an issue of Odoo, I made the following PR https://github.com/odoo/odoo/pull/108609 to fix this issue.

I post this in case you face the same issue 😄

**Pull request merged**